File: src/lib/storage/fs_management/cpp/format.cc
// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/lib/storage/fs_management/cpp/format.h"

#include <fidl/fuchsia.hardware.block/cpp/wire.h>
#include <lib/fdio/cpp/caller.h>
#include <unistd.h>

#include <mutex>
#include <unordered_map>

#include <fbl/algorithm.h>
#include <pretty/hexdump.h>

#include "src/lib/fxl/synchronization/thread_annotations.h"
#include "src/lib/storage/block_client/cpp/remote_block_device.h"
#include "src/lib/storage/fs_management/cpp/path.h"

namespace fs_management {
namespace {

namespace fblock = fuchsia_hardware_block;

class Registry {
 public:
  DiskFormat Register(std::unique_ptr<CustomDiskFormat> format) {
    std::scoped_lock lock(mutex_);
    int format_id = next_id_++;
    map_[format_id] = std::move(format);
    return static_cast<DiskFormat>(format_id);
  }

  const CustomDiskFormat* Get(DiskFormat format) {
    std::scoped_lock lock(mutex_);
    auto iter = map_.find(format);
    if (iter == map_.end()) {
      return nullptr;
    }
    return iter->second.get();
  }

 private:
  std::mutex mutex_;
  int next_id_ FXL_GUARDED_BY(mutex_) = kDiskFormatCount;
  std::unordered_map<int, std::unique_ptr<CustomDiskFormat>> map_ FXL_GUARDED_BY(mutex_);
};

Registry& GetRegistry() {
  static Registry& registry = *new Registry;
  return registry;
}

enum DiskFormatLogVerbosity {
  Silent,
  Verbose,
};

DiskFormat DetectDiskFormatImpl(fidl::UnownedClientEnd<fblock::Block> device,
                                DiskFormatLogVerbosity verbosity) {
  const fidl::WireResult result = fidl::WireCall(device)->GetInfo();
  if (!result.ok()) {
    fprintf(stderr, "DetectDiskFormat: Could not acquire block device info: %s\n",
            result.FormatDescription().c_str());
    return kDiskFormatUnknown;
  }
  const fit::result response = result.value();
  if (response.is_error()) {
    fprintf(stderr, "DetectDiskFormat: Could not acquire block device info: %s\n",
            zx_status_get_string(response.error_value()));
    return kDiskFormatUnknown;
  }

  const fuchsia_hardware_block::wire::BlockInfo& info = response.value()->info;
  if (info.block_size == 0) {
    fprintf(stderr, "DetectDiskFormat: Expected a block size of > 0\n");
    return kDiskFormatUnknown;
  }

  // We need to read at least two blocks, because the GPT magic is located inside the second block
  // of the disk.
  size_t header_size = (kHeaderSize > (2 * info.block_size)) ? kHeaderSize : (2 * info.block_size);
  // check if the partition is big enough to hold the header in the first place
  if (header_size > info.block_size * info.block_count) {
    return kDiskFormatUnknown;
  }

  // We expect to read kHeaderSize bytes, but we may need to read
  // extra to read a multiple of the underlying block size.
  const size_t buffer_size = fbl::round_up(header_size, static_cast<size_t>(info.block_size));

  ZX_DEBUG_ASSERT_MSG(buffer_size > 0, "Expected buffer_size to be greater than 0\n");

  uint8_t data[buffer_size];
  {
    auto result = block_client::SingleReadBytes(device, data, buffer_size, 0);
    if (result != ZX_OK) {
      fprintf(stderr, "DetectDiskFormat: Error reading block device.\n");
      return kDiskFormatUnknown;
    }
  }

  if (!memcmp(data, kFvmMagic, sizeof(kFvmMagic))) {
    return kDiskFormatFvm;
  }

  if (!memcmp(data, kZxcryptMagic, sizeof(kZxcryptMagic))) {
    return kDiskFormatZxcrypt;
  }

  if (!memcmp(data, kBlockVerityMagic, sizeof(kBlockVerityMagic))) {
    return kDiskFormatBlockVerity;
  }

  if (!memcmp(data + info.block_size, kGptMagic, sizeof(kGptMagic))) {
    return kDiskFormatGpt;
  }

  if (!memcmp(data, kMinfsMagic, sizeof(kMinfsMagic))) {
    return kDiskFormatMinfs;
  }

  if (!memcmp(data, kBlobfsMagic, sizeof(kBlobfsMagic))) {
    return kDiskFormatBlobfs;
  }

  if (!memcmp(data, kFactoryfsMagic, sizeof(kFactoryfsMagic))) {
    return kDiskFormatFactoryfs;
  }

  if (!memcmp(data, kVbmetaMagic, sizeof(kVbmetaMagic))) {
    return kDiskFormatVbmeta;
  }

  if ((data[510] == 0x55 && data[511] == 0xAA)) {
    if ((data[38] == 0x29 || data[66] == 0x29)) {
      // 0x55AA are always placed at offset 510 and 511 for FAT filesystems.
      // 0x29 is the Boot Signature, but it is placed at either offset 38 or
      // 66 (depending on FAT type).
      return kDiskFormatFat;
    }
    return kDiskFormatMbr;
  }

  if (!memcmp(&data[1024], kF2fsMagic, sizeof(kF2fsMagic))) {
    return kDiskFormatF2fs;
  }

  if (!memcmp(data, kFxfsMagic, sizeof(kFxfsMagic))) {
    return kDiskFormatFxfs;
  }

  if (verbosity == DiskFormatLogVerbosity::Verbose) {
    // Log a hexdump of the bytes we looked at and didn't find any magic in.
    fprintf(stderr, "DetectDiskFormat: did not recognize format.  Looked at:\n");
    // fvm, zxcrypt, minfs, and blobfs have their magic bytes at the start
    // of the block.
    hexdump_very_ex(data, 16, 0, hexdump_stdio_printf, stderr);
    // MBR is two bytes at offset 0x1fe, but print 16 just for consistency
    hexdump_very_ex(data + 0x1f0, 16, 0x1f0, hexdump_stdio_printf, stderr);
    // GPT magic is stored one block in, so it can coexist with MBR.
    hexdump_very_ex(data + info.block_size, 16, info.block_size, hexdump_stdio_printf, stderr);
  }

  return kDiskFormatUnknown;
}

}  // namespace

__EXPORT
DiskFormat DetectDiskFormat(fidl::UnownedClientEnd<fblock::Block> device) {
  return DetectDiskFormatImpl(device, DiskFormatLogVerbosity::Silent);
}

__EXPORT
DiskFormat DetectDiskFormatLogUnknown(fidl::UnownedClientEnd<fblock::Block> device) {
  return DetectDiskFormatImpl(device, DiskFormatLogVerbosity::Verbose);
}

DiskFormat CustomDiskFormat::Register(std::unique_ptr<CustomDiskFormat> format) {
  return GetRegistry().Register(std::move(format));
}

const CustomDiskFormat* CustomDiskFormat::Get(DiskFormat format) {
  return GetRegistry().Get(format);
}

__EXPORT std::string_view DiskFormatString(DiskFormat fs_type) {
  switch (fs_type) {
    case kDiskFormatCount:
    case kDiskFormatUnknown:
      break;
    case kDiskFormatGpt:
      return "gpt";
    case kDiskFormatMbr:
      return "mbr";
    case kDiskFormatMinfs:
      return "minfs";
    case kDiskFormatFat:
      return "fat";
    case kDiskFormatBlobfs:
      return "blobfs";
    case kDiskFormatFvm:
      return "fvm";
    case kDiskFormatZxcrypt:
      return "zxcrypt";
    case kDiskFormatFactoryfs:
      return "factoryfs";
    case kDiskFormatBlockVerity:
      return "block verity";
    case kDiskFormatVbmeta:
      return "vbmeta";
    case kDiskFormatBootpart:
      return "bootpart";
    case kDiskFormatFxfs:
      return "fxfs";
    case kDiskFormatF2fs:
      return "f2fs";
    case kDiskFormatNandBroker:
      return "nand broker";
  }

  auto format = CustomDiskFormat::Get(fs_type);
  if (format == nullptr) {
    return "unknown!";
  }

  return format->name().c_str();
}

__EXPORT DiskFormat DiskFormatFromString(std::string_view str) {
  static auto* formats = [] {
    auto* formats = new std::unordered_map<std::string_view, DiskFormat>();
    for (auto format : {kDiskFormatGpt, kDiskFormatMbr, kDiskFormatMinfs, kDiskFormatFat,
                        kDiskFormatBlobfs, kDiskFormatFvm, kDiskFormatZxcrypt, kDiskFormatFactoryfs,
                        kDiskFormatBlockVerity, kDiskFormatVbmeta, kDiskFormatBootpart,
                        kDiskFormatFxfs, kDiskFormatF2fs, kDiskFormatNandBroker}) {
      formats->emplace(DiskFormatString(format), format);
    }
    return formats;
  }();
  auto iter = formats->find(str);
  if (iter == formats->end()) {
    return kDiskFormatUnknown;
  }
  return iter->second;
}

__EXPORT std::string_view DiskFormatComponentUrl(DiskFormat fs_type) {
  switch (fs_type) {
    case kDiskFormatBlobfs:
      return kBlobfsComponentUrl;
    case kDiskFormatFxfs:
      return kFxfsComponentUrl;
    case kDiskFormatMinfs:
      return kMinfsComponentUrl;
    case kDiskFormatF2fs:
      return kF2fsComponentUrl;
    case kDiskFormatCount:
    case kDiskFormatUnknown:
    case kDiskFormatGpt:
    case kDiskFormatMbr:
    case kDiskFormatFat:
    case kDiskFormatFvm:
    case kDiskFormatZxcrypt:
    case kDiskFormatFactoryfs:
    case kDiskFormatBlockVerity:
    case kDiskFormatVbmeta:
    case kDiskFormatBootpart:
    case kDiskFormatNandBroker:
      break;
  }

  auto format = CustomDiskFormat::Get(fs_type);
  if (format == nullptr) {
    return "";
  }

  return format->url().c_str();
}

__EXPORT std::string DiskFormatBinaryPath(DiskFormat fs_type) {
  switch (fs_type) {
    case kDiskFormatBlobfs:
      return GetBinaryPath("blobfs");
    case kDiskFormatMinfs:
      return GetBinaryPath("minfs");
    case kDiskFormatF2fs:
      return GetBinaryPath("f2fs");
    case kDiskFormatFat:
      return GetBinaryPath("fatfs");
    case kDiskFormatFactoryfs:
      return GetBinaryPath("factoryfs");
    case kDiskFormatFxfs:
    case kDiskFormatCount:
    case kDiskFormatUnknown:
    case kDiskFormatGpt:
    case kDiskFormatMbr:
    case kDiskFormatFvm:
    case kDiskFormatZxcrypt:
    case kDiskFormatBlockVerity:
    case kDiskFormatVbmeta:
    case kDiskFormatBootpart:
    case kDiskFormatNandBroker:
      break;
  }

  auto format = CustomDiskFormat::Get(fs_type);
  if (format == nullptr) {
    return "";
  }
  return format->binary_path();
}

}  // namespace fs_management

File: src/lib/storage/fs_management/rust/src/format.rs
// Copyright 2022 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

pub mod constants;

use {
    anyhow::{anyhow, ensure, Context, Error},
    fidl_fuchsia_hardware_block::BlockProxy,
    fuchsia_zircon::{self as zx, HandleBased},
};

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum DiskFormat {
    Unknown,
    Gpt,
    Mbr,
    Minfs,
    Fat,
    Blobfs,
    Fvm,
    Zxcrypt,
    FactoryFs,
    BlockVerity,
    VbMeta,
    BootPart,
    Fxfs,
    F2fs,
    NandBroker,
}

pub fn round_up(val: u64, divisor: u64) -> u64 {
    ((val + (divisor - 1)) / divisor) * divisor
}

pub async fn detect_disk_format(block_proxy: &BlockProxy) -> DiskFormat {
    match detect_disk_format_res(&block_proxy).await {
        Ok(format) => format,
        Err(e) => {
            tracing::error!("detect_disk_format failed: {}", e);
            return DiskFormat::Unknown;
        }
    }
}

async fn detect_disk_format_res(block_proxy: &BlockProxy) -> Result<DiskFormat, Error> {
    let (status, info) =
        block_proxy.get_info().await.context("transport error on get_info call")?;
    zx::Status::ok(status).context("get_info call failed")?;
    let block_info = info.ok_or(anyhow!("Expected BlockInfo"))?;

    ensure!(block_info.block_size > 0, "block size expected to be non-zero");

    let header_size = if constants::HEADER_SIZE > 2 * block_info.block_size {
        constants::HEADER_SIZE
    } else {
        2 * block_info.block_size
    };

    ensure!(
        header_size <= block_info.block_size * block_info.block_count as u32,
        "header size is larger than the size of the block device"
    );

    let buffer_size = round_up(header_size as u64, block_info.block_size as u64);
    let vmo = zx::Vmo::create(buffer_size).context("failed to create vmo")?;
    let status = block_proxy
        .read_blocks(
            vmo.duplicate_handle(zx::Rights::SAME_RIGHTS)
                .context("vmo duplicate handle call failed")?,
            buffer_size,
            0,
            0,
        )
        .await
        .context("transport error on read_blocks protocol")?;

    zx::Status::ok(status).context("read_blocks returned an error status")?;

    let mut data = vec![0; buffer_size as usize];
    vmo.read(&mut data, 0).context("vmo read failed")?;

    if data.starts_with(&constants::FVM_MAGIC) {
        return Ok(DiskFormat::Fvm);
    }

    if data.starts_with(&constants::ZXCRYPT_MAGIC) {
        return Ok(DiskFormat::Zxcrypt);
    }

    if data.starts_with(&constants::BLOCK_VERITY_MAGIC) {
        return Ok(DiskFormat::BlockVerity);
    }

    if &data[block_info.block_size as usize..block_info.block_size as usize + 16]
        == &constants::GPT_MAGIC
    {
        return Ok(DiskFormat::Gpt);
    }

    if data.starts_with(&constants::MINFS_MAGIC) {
        return Ok(DiskFormat::Minfs);
    }

    if data.starts_with(&constants::BLOBFS_MAGIC) {
        return Ok(DiskFormat::Blobfs);
    }

    if data.starts_with(&constants::FACTORYFS_MAGIC) {
        return Ok(DiskFormat::FactoryFs);
    }

    if data.starts_with(&constants::VB_META_MAGIC) {
        return Ok(DiskFormat::VbMeta);
    }

    if data[510] == 0x55 && data[511] == 0xAA {
        if data[38] == 0x29 || data[66] == 0x29 {
            // 0x55AA are always placed at offset 510 and 511 for FAT filesystems.
            // 0x29 is the Boot Signature, but it is placed at either offset 38 or
            // 66 (depending on FAT type).
            return Ok(DiskFormat::Fat);
        }
        return Ok(DiskFormat::Mbr);
    }

    if &data[1024..1024 + constants::F2FS_MAGIC.len()] == &constants::F2FS_MAGIC {
        return Ok(DiskFormat::F2fs);
    }

    if data.starts_with(&constants::FXFS_MAGIC) {
        return Ok(DiskFormat::Fxfs);
    }

    return Ok(DiskFormat::Unknown);
}

#[cfg(test)]
mod tests {
    use {
        super::{constants, detect_disk_format, DiskFormat},
        fidl::endpoints::create_proxy_and_stream,
        fidl_fuchsia_hardware_block::{BlockInfo, BlockMarker, BlockRequest},
        fuchsia_zircon as zx,
        futures::{pin_mut, select, FutureExt, TryStreamExt},
    };

    async fn get_detected_disk_format(content: &[u8]) -> DiskFormat {
        let (proxy, mut stream) = create_proxy_and_stream::<BlockMarker>().unwrap();

        let mock_device = async {
            while let Some(request) = stream.try_next().await.unwrap() {
                match request {
                    BlockRequest::GetInfo { responder } => {
                        responder
                            .send(
                                zx::sys::ZX_OK,
                                Some(&mut BlockInfo {
                                    block_count: 1000,
                                    block_size: 512,
                                    max_transfer_size: 1024 * 1024,
                                    flags: 0,
                                    reserved: 0,
                                }),
                            )
                            .unwrap();
                    }
                    BlockRequest::ReadBlocks { vmo, length, dev_offset, vmo_offset, responder } => {
                        assert_eq!(dev_offset, 0);
                        assert_eq!(length, 4096);
                        vmo.write(content, vmo_offset).unwrap();
                        responder.send(zx::sys::ZX_OK).unwrap();
                    }
                    _ => unreachable!(),
                }
            }
        }
        .fuse();

        pin_mut!(mock_device);

        select! {
          _ = mock_device => unreachable!(),
          format = detect_disk_format(&proxy).fuse() => format,
        }
    }

    #[fuchsia::test]
    async fn detect_format_fvm() {
        let mut data = vec![0; 4096];
        data[..constants::FVM_MAGIC.len()].copy_from_slice(&constants::FVM_MAGIC);
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::Fvm);
    }

    #[fuchsia::test]
    async fn detect_format_zxcrypt() {
        let mut data = vec![0; 4096];
        data[0..constants::ZXCRYPT_MAGIC.len()].copy_from_slice(&constants::ZXCRYPT_MAGIC);
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::Zxcrypt);
    }

    #[fuchsia::test]
    async fn detect_format_block_verity() {
        let mut data = vec![0; 4096];
        data[0..constants::BLOCK_VERITY_MAGIC.len()]
            .copy_from_slice(&constants::BLOCK_VERITY_MAGIC);
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::BlockVerity);
    }

    #[fuchsia::test]
    async fn detect_format_gpt() {
        let mut data = vec![0; 4096];
        data[512..512 + 16].copy_from_slice(&constants::GPT_MAGIC);
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::Gpt);
    }

    #[fuchsia::test]
    async fn detect_format_minfs() {
        let mut data = vec![0; 4096];
        data[0..constants::MINFS_MAGIC.len()].copy_from_slice(&constants::MINFS_MAGIC);
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::Minfs);
    }

    #[fuchsia::test]
    async fn detect_format_blobfs() {
        let mut data = vec![0; 4096];
        data[0..constants::BLOBFS_MAGIC.len()].copy_from_slice(&constants::BLOBFS_MAGIC);
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::Blobfs);
    }

    #[fuchsia::test]
    async fn detect_format_factory_fs() {
        let mut data = vec![0; 4096];
        data[0..constants::FACTORYFS_MAGIC.len()].copy_from_slice(&constants::FACTORYFS_MAGIC);
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::FactoryFs);
    }

    #[fuchsia::test]
    async fn detect_format_vb_meta() {
        let mut data = vec![0; 4096];
        data[0..constants::VB_META_MAGIC.len()].copy_from_slice(&constants::VB_META_MAGIC);
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::VbMeta);
    }

    #[fuchsia::test]
    async fn detect_format_mbr() {
        let mut data = vec![0; 4096];
        data[510] = 0x55 as u8;
        data[511] = 0xAA as u8;
        data[38] = 0x30 as u8;
        data[66] = 0x30 as u8;
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::Mbr);
    }

    #[fuchsia::test]
    async fn detect_format_fat_1() {
        let mut data = vec![0; 4096];
        data[510] = 0x55 as u8;
        data[511] = 0xAA as u8;
        data[38] = 0x29 as u8;
        data[66] = 0x30 as u8;
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::Fat);
    }

    #[fuchsia::test]
    async fn detect_format_fat_2() {
        let mut data = vec![0; 4096];
        data[510] = 0x55 as u8;
        data[511] = 0xAA as u8;
        data[38] = 0x30 as u8;
        data[66] = 0x29 as u8;
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::Fat);
    }

    #[fuchsia::test]
    async fn detect_format_f2fs() {
        let mut data = vec![0; 4096];
        data[1024..1024 + constants::F2FS_MAGIC.len()].copy_from_slice(&constants::F2FS_MAGIC);
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::F2fs);
    }

    #[fuchsia::test]
    async fn detect_format_fxfs() {
        let mut data = vec![0; 4096];
        data[0..constants::FXFS_MAGIC.len()].copy_from_slice(&constants::FXFS_MAGIC);
        assert_eq!(get_detected_disk_format(&data).await, DiskFormat::Fxfs);
    }

    #[fuchsia::test]
    async fn detect_format_unknown() {
        assert_eq!(get_detected_disk_format(&vec![0; 4096]).await, DiskFormat::Unknown);
    }
}
